tests/integration/pilot/upgrade/multi_version_revision_test.go

```diff
@@ -0,0 +1,167 @@
+// +build integ
+// Copyright Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package upgrade
+
+import (
+	"fmt"
+	"io/ioutil"
+	"path/filepath"
+	"testing"
+	"time"
+
+	"istio.io/istio/pkg/config/protocol"
+	"istio.io/istio/pkg/test/env"
+	"istio.io/istio/pkg/test/framework"
+	"istio.io/istio/pkg/test/framework/components/echo"
+	"istio.io/istio/pkg/test/framework/components/echo/echoboot"
+	"istio.io/istio/pkg/test/framework/components/istio"
+	"istio.io/istio/pkg/test/framework/components/namespace"
+	"istio.io/istio/pkg/test/util/retry"
+)
+
+var i istio.Instance
+
+// TestMain defines the entrypoint for pilot tests using a standard Istio installation.
+// If a test requires a custom install it should go into its own package, otherwise it should go
+// here to reuse a single install across tests.
+func TestMain(m *testing.M) {
+	framework.
+		NewSuite(m).
+		RequireSingleCluster().
+		Setup(istio.Setup(&i, nil)).
+		Run()
+}
+
+type installGolden struct {
+	revision string
+	path     string
+}
+
+type revisionedNamespace struct {
+	revision  string
+	namespace namespace.Instance
+}
+
+// TestMultiVersionRevision tests traffic between data planes running under differently versioned revisions
+// should test all possible revisioned namespace pairings to test traffic between all versions
+func TestMultiVersionRevision(t *testing.T) {
+	framework.NewTest(t).
+		RequiresSingleCluster().
+		Features("installation.upgrade").
+		Run(func(ctx framework.TestContext) {
+			goldenDir := filepath.Join(env.IstioSrc, "tests/integration/pilot/upgrade/goldens")
```

Comment:
nit: goldens -> testdata

Aside from standard practice, I think golang treats this folder slightly differently

Reply:
Changed